samples/boards/nrf52/mesh/onoff-app/src/main.c

```diff
@@ -0,0 +1,749 @@
+};
+
+static struct bt_mesh_health_srv health_cli = {
+};
+
+/*
+ * Publication Declarations
+ *
+ * The publication messages are initialized to the
+ * the size of the opcode + content + MIC
+ *
+ * The messages are in static storage because NET_BUF_SIMPLE()
+ * only allocates on the stack if called within a function.
+ * For publication, the message must be in static or global as
+ * it is re-transmitted several times. This occurs
+ * after the function that called bt_mesh_model_publish() has
+ * exited and the stack is no longer valid.
+ *
+ */
+
+static struct bt_mesh_model_pub health_pub = {
+	.msg  = BT_MESH_HEALTH_FAULT_MSG(0),
+};
+
+static struct bt_mesh_model_pub gen_level_pub;
+static struct bt_mesh_model_pub gen_onoff_pub_srv = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_srv_s_0 = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli_s_0 = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_srv_s_1 = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli_s_1 = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_srv_s_2 = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli_s_2 = {
+	.msg = NET_BUF_SIMPLE(2 + 2 + 4),
+};
+
+struct device *led_device[4], *sw_device;
+
+struct sw {
+	u32_t pin;
+	struct k_work button_work;
+};
+
+static struct sw sw;
+
+static struct gpio_callback button_cb;
+
+static int led_gpio_pin[] = {
+	LED0_GPIO_PIN,
+	LED1_GPIO_PIN,
+	LED2_GPIO_PIN,
+	LED3_GPIO_PIN,
+};
+
+static u8_t onoff_state[4] = {0};
+static u8_t last_onoff_state[4] = {1, 1, 1, 1};
+static u8_t level;
+static u8_t trans_id;
+static u32_t time, last_time;
+static u16_t primary_addr;
+static u16_t primary_net_idx;
+
+/*
+ * Generic OnOff Model Server Message Handlers
+ *
+ * Mesh Model Specification 3.1.1
+ *
+ */
+
+static void gen_onoff_get(struct bt_mesh_model *model,
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	struct net_buf_simple *msg = NET_BUF_SIMPLE(2 + 1 + 4);
+	u8_t elem_idx;
+
+	BT_DBG("");
+	elem_idx = model->elem->addr - primary_addr;
+	bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_GEN_ONOFF_STATUS);
+	net_buf_simple_add_u8(msg, onoff_state[elem_idx]);
+
+	if (bt_mesh_model_send(model, ctx, msg, NULL, NULL)) {
+		BT_ERR("Unable to send On Off Status response");
+	}
+}
+
+static void gen_onoff_set_unack(struct bt_mesh_model *model,
+				struct bt_mesh_msg_ctx *ctx,
+				struct net_buf_simple *buf)
+{
+	struct net_buf_simple *msg = model->pub->msg;
+	u8_t elem_idx = model->elem->addr - primary_addr;
+	int err;
+	u8_t local_trans_id;
+
+	/* Ignore command sent by us to us as the LED state is set by button */
+	if (model->elem->addr != ctx->addr) {
```

Comment:
Wouldn't it make more sense to do:

```
if (model->elem->addr == ctx->addr) {
        return;
}
````

It's a bit unclear why you need any extra processing in the case of a message from yourself.

Reply:
The issue here is that the button changes the led state directly. I did this so the pair act like a switch prior to provisioning. Consequently, when the button broadcasts to the same group that its led subscribes, the led server will also receive the message and change its state again. So, I ignore messages sent from its button.

I also need to execute the to execute the code block that checks if the led server needs to publish on a state change.

I'm open to any suggestion of a cleaner way to do this.

Reply:
I must not have got enough sleep. I now understand your comment. I don't have to consider the publish, it was done by the button.

Reply:
fixed